src/coreclr/jit/gentree.cpp

```diff
@@ -2724,48 +2724,48 @@ bool GenTree::Compare(GenTree* op1, GenTree* op2, bool swapOK)
     return result;
 }
 
-struct AddrTakenDsc
-{
-    Compiler* comp;
-    bool      hasAddrTakenLcl;
-};
-
-/* static */
-Compiler::fgWalkResult Compiler::gtHasLocalsWithAddrOpCB(GenTree** pTree, fgWalkData* data)
+//------------------------------------------------------------------------------
+// gtHasLocalsWithAddrOp:
+//   Check if this tree contains locals with lvHasLdAddrOp or
+//   IsAddressExposed() flags set. Does a full tree walk.
+//
+// Paramters:
+//   tree - the tree
+//
+// Return Value:
+//    True if any sub tree is such a local.
+//
+bool Compiler::gtHasLocalsWithAddrOp(GenTree* tree)
 {
-    GenTree*  tree = *pTree;
-    Compiler* comp = data->compiler;
-
-    if (tree->gtOper == GT_LCL_VAR)
+    struct LocalsWithAddrOpVisitor : GenTreeVisitor<LocalsWithAddrOpVisitor>
     {
-        const LclVarDsc* varDsc = comp->lvaGetDesc(tree->AsLclVarCommon());
-
-        if (varDsc->lvHasLdAddrOp || varDsc->IsAddressExposed())
+        enum
         {
-            ((AddrTakenDsc*)data->pCallbackData)->hasAddrTakenLcl = true;
-            return WALK_ABORT;
-        }
-    }
-
-    return WALK_CONTINUE;
-}
+            DoPreOrder    = true,
+            DoLclVarsOnly = true,
+        };
 
-/*****************************************************************************
- *
- *  Return true if this tree contains locals with lvHasLdAddrOp or IsAddressExposed()
- *  flag(s) set.
- */
+        bool HasAddrTakenLocal = false;
 
-bool Compiler::gtHasLocalsWithAddrOp(GenTree* tree)
-{
-    AddrTakenDsc desc;
+        LocalsWithAddrOpVisitor(Compiler* comp) : GenTreeVisitor(comp)
+        {
+        }
 
-    desc.comp            = this;
-    desc.hasAddrTakenLcl = false;
+        fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
+        {
+            LclVarDsc* varDsc = m_compiler->lvaGetDesc((*use)->AsLclVarCommon());
+            if (varDsc->lvHasLdAddrOp || varDsc->IsAddressExposed())
+            {
+                HasAddrTakenLocal = true;
```

Comment:
Shouldn't we still early out like before?
```suggestion
                HasAddrTakenLocal = true;
                return WALK_ABORT;
```

Reply:
Whoops, yes, good catch.